spec.bs

```diff
@@ -201,56 +201,242 @@ This document introduces a new storage API that is intentionally not partitioned
     1. [=Assert=] that |globalObject| is a {{Window}} or a {{SharedStorageWorkletGlobalScope}}.
-    1. If |globalObject| is a {{Window}} and |globalObject|'s [=associated document=] is not [=allowed to use=] the "[=PermissionsPolicy/shared-storage=]" feature, return false.
+    1. If |globalObject| is a {{Window}}, and if the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage=]", |globalObject|'s [=associated document=], and |origin| returns false, then return false.
+    1. Return true.
+  </div>
+
+  <div algorithm>
+    To <dfn>determine whether shared storage is allowed by enrollment and user preference</dfn>, given an [=environment settings object=] |environment| and an [=url/origin=] |origin|, run these steps:
+
     1. If the result of running [=obtaining a site|obtain a site=] with |origin| is not [=enrolled=], then return false.
-    1. If the result of running [=check if user preference setting allows access to shared storage=] from |environment| is false, then return false.
+    1. If the result of running [=check if user preference setting allows access to shared storage=] given |environment| and |origin| is false, then return false.
     1. Return true.
   </div>
 
+  <div class="note">
+    Here are the scenarios where the algorithms [=determine whether shared storage is allowed by context=] and [=determine whether shared storage is allowed by enrollment and user preference=] could be used:
+
+    - For <a href="#window-setter">Window setter/deleter methods</a>, |environment|'s [=environment settings object/origin=] should equal |origin|.
+    - For creating a worklet, and for initiating (from {{Window}}) and running (from {{SharedStorageWorkletGlobalScope}}) operations on a worklet, |environment| should be the worklet's creator window, and |origin| should be the [=SharedStorageWorklet/worklet origin=].
+    - For a <a href="#ss-fetch-algo">fetch request</a>, |environment| should be the request's initiator window, and |origin| should be the request's URL's origin.
+  </div>
+
   <div algorithm>
-    To <dfn>check if addModule is allowed and update status</dfn> for a {{SharedStorageWorklet}} |worklet|, run the following steps:
-    1. If the result of running [=determine whether shared storage is allowed=] on the [=relevant settings object=] of [=this=] is false, return false.
-    1. If |worklet|'s [=addModule initiated=] is true, return false.
+    To <dfn>check if addModule is allowed and update status</dfn> given a {{SharedStorageWorklet}} |worklet| and an [=/URL=] |moduleURL|, run the following steps:
+    1. Let |workletOrigin| be |moduleURL|'s [=url/origin=].
+    1. If the result of running [=determine whether shared storage is allowed by context=] given [=this=]'s [=relevant settings object=] and |workletOrigin| is false, return "DisallowedDueToNonPreferenceError".
+    1. If |worklet|'s [=addModule initiated=] is true, return "DisallowedDueToNonPreferenceError".
```

Comment:
It can be set if the algorithm (addModule()) was invoked on the same worklet object. Yes, the [=SharedStorageWorklet/worklet origin=] could also be set at this point.

Reply:
`|workletOrigin|` could be null here though, which breaks the call to "determine whether SS is allowed by context", which expects an origin (not null). So can you:

1. Only conditionally call this if `workletOrigin` is non-null
2. In the case where it is non-null, assert that it equals the [worklet origin member](https://pr-preview.s3.amazonaws.com/WICG/shared-storage/pull/131.html#sharedstorageworklet-worklet-origin)
    - Or would this assert be invalid? I guess maybe it is invalid because you can call `addModule()` several times with different (cross-origin) URLs?

?

Reply:
Ping on this.

Reply:
I just saw https://github.com/WICG/shared-storage/pull/131#issuecomment-1972259132. I'll respond there.